src/pools/D3M4626TypePool.sol

```diff
@@ -0,0 +1,153 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity ^0.8.14;
+
+import "./ID3MPool.sol";
+import "forge-std/interfaces/IERC4626.sol";
+
+interface VatLike {
+    function live() external view returns (uint256);
+    function hope(address) external;
+    function nope(address) external;
+}
+
+interface D3mHubLike {
+    function vat() external view returns (address);
+    function end() external view returns (EndLike);
+}
+
+interface EndLike {
+    function Art(bytes32) external view returns (uint256);
+}
+
+contract D3M4626TypePool is ID3MPool {
+
+    mapping (address => uint256) public wards;
+    address                      public hub;
+    uint256                      public exited;
+
+    bytes32  public immutable ilk;
+    VatLike  public immutable vat;
+    IERC4626 public immutable vault;
+    IERC20   public immutable dai;
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event File(bytes32 indexed what, address data);
+
+    constructor(bytes32 ilk_, address hub_, address dai_, address vault_) {
+        ilk = ilk_;
+        dai = IERC20(dai_);
+        vault = IERC4626(vault_);
+
+        require(ilk_ != bytes32(0), "D3M4626TypePool/zero-bytes32");
+        require(hub_ != address(0), "D3M4626TypePool/zero-address");
+        require(dai_ != address(0), "D3M4626TypePool/zero-address");
+        require(vault_ != address(0), "D3M4626TypePool/zero-address");
+        require(IERC4626(vault_).asset() == dai_, "D3M4626TypePool/vault-asset-is-not-dai");
+
+        dai.approve(vault_, type(uint256).max);
+
+        hub = hub_;
+        vat = VatLike(D3mHubLike(hub).vat());
+        vat.hope(hub_);
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    modifier onlyHub() {
+        require(msg.sender == hub, "D3M4626TypePool/only-hub");
+        _;
+    }
+
+    modifier auth() {
+        require(wards[msg.sender] == 1, "D3M4626TypePool/not-authorized");
+        _;
+    }
+
+    // --- Admin ---
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    function file(bytes32 what, address data) external auth {
+        require(vat.live() == 1, "D3M4626TypePool/no-file-during-shutdown");
+        if (what == "hub") {
+            vat.nope(hub);
+            hub = data;
+            vat.hope(data);
+        } else revert("D3M4626TypePool/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    /// @inheritdoc ID3MPool
+    function deposit(uint256 wad) external override onlyHub {
```

Comment:
It probably helps adding in a comment the link to the function:
https://github.com/morpho-org/metamorpho/blob/fcf3c41d9c113514c9af0bbf6298e88a1060b220/src/MetaMorpho.sol#L531

Reply:
done